chain-tx-enclave/enclave-ra/ra-client/src/verifier.rs

```diff
@@ -0,0 +1,226 @@
+        let mut attestation_report_received = false;
+
+        let public_key_hash = sha256(
+            certificate
+                .tbs_certificate
+                .subject_pki
+                .subject_public_key
+                .data,
+        );
+
+        for extension in certificate.tbs_certificate.extensions {
+            if extension.oid == attestation_report_oid {
+                attestation_report_received = true;
+                self.verify_attestation_report(extension.value, public_key_hash)?;
+            }
+        }
+
+        if attestation_report_received {
+            Ok(())
+        } else {
+            Err(EnclaveCertVerifierError::MissingAttestationReport)
+        }
+    }
+
+    /// Verifies attestation report
+    fn verify_attestation_report(
+        &self,
+        attestation_report: &[u8],
+        public_key_hash: [u8; 32],
+    ) -> Result<(), EnclaveCertVerifierError> {
+        log::info!("Verifying attestation report");
+
+        let trust_anchors: Vec<TrustAnchor> = self
+            .root_cert_store
+            .roots
+            .iter()
+            .map(|cert| cert.to_trust_anchor())
+            .collect();
+        let time =
+            Time::try_from(SystemTime::now()).map_err(|_| EnclaveCertVerifierError::TimeError)?;
+
+        let attestation_report: AttestationReport = bincode::deserialize(attestation_report)?;
+
+        let signing_certs = certs(&mut attestation_report.signing_cert.as_ref())
+            .map_err(|_| EnclaveCertVerifierError::CertificateParsingError)?;
+
+        for signing_cert in signing_certs {
+            let signing_cert = EndEntityCert::from(&signing_cert.0)?;
+
+            signing_cert.verify_is_valid_tls_server_cert(
+                SUPPORTED_SIG_ALGS,
+                &TLSServerTrustAnchors(&trust_anchors),
+                &[],
+                time,
+            )?;
+
+            signing_cert.verify_signature(
+                &RSA_PKCS1_2048_8192_SHA256,
+                &attestation_report.body,
+                &attestation_report.signature,
+            )?;
+        }
+
+        self.verify_attestation_report_body(&attestation_report.body, public_key_hash)?;
+
+        log::info!("Attestation report is valid!");
+        Ok(())
+    }
+
+    fn verify_attestation_report_body(
+        &self,
+        attestation_report_body: &[u8],
+        public_key_hash: [u8; 32],
+    ) -> Result<(), EnclaveCertVerifierError> {
+        let attestation_report_body: AttestationReportBody =
+            serde_json::from_slice(attestation_report_body)?;
+
+        if !self
+            .valid_enclave_quote_statuses
+            .contains(&attestation_report_body.isv_enclave_quote_status)
+        {
+            return Err(EnclaveCertVerifierError::InvalidEnclaveQuoteStatus(
+                attestation_report_body.isv_enclave_quote_status,
+            ));
+        }
+
+        let quote = attestation_report_body.get_quote()?;
```

Comment:
In enclave, a new certificate (including quote and attesation report) is created after every `EnclaveRaContext.validity_duration` seconds. In [this](https://github.com/crypto-com/chain/pull/1311/files#diff-e5b664cc677a2e0aa8c2891c1142cf24) file, we mention that:

>If the certificate generation fails, we do not crash and keep using the old certificate (if available) and client can decide if they want to use the old certificate. Every certificate has a 90 days valid duration. If certificate creation fails for 90 days, the enclave itself will not serve any client.

In client code, we do check the validity of certificate (i.e., 90 days). Maybe add a `cert_validity_duration` in `ra-client` too?

Reply:
well, but the quote itself needs to be checked as well -- it can't just be "trusted"... otherwise one can just keep recreating new certs with some leaked key and reusing an old quote

Reply:
Done.